crates/bevy_asset/src/loader.rs

```diff
@@ -1,274 +1,397 @@
-    pub(crate) asset_io: &'a dyn AssetIo,
-    pub(crate) labeled_assets: HashMap<Option<String>, BoxedLoadedAsset>,
-    pub(crate) path: &'a Path,
-    pub(crate) version: usize,
+    asset_server: &'a AssetServer,
+    should_load_dependencies: bool,
+    asset_path: AssetPath<'static>,
+    dependencies: HashSet<UntypedHandle>,
+    /// Direct dependencies used by this loader.
+    loader_dependencies: HashMap<AssetPath<'static>, u64>,
+    labeled_assets: HashMap<String, LabeledAsset>,
 }
 
 impl<'a> LoadContext<'a> {
     pub(crate) fn new(
-        path: &'a Path,
-        ref_change_channel: &'a RefChangeChannel,
-        asset_io: &'a dyn AssetIo,
-        version: usize,
+        asset_server: &'a AssetServer,
+        asset_path: AssetPath<'static>,
+        load_dependencies: bool,
     ) -> Self {
         Self {
-            ref_change_channel,
-            asset_io,
-            labeled_assets: Default::default(),
-            version,
-            path,
+            asset_server,
+            asset_path,
+            should_load_dependencies: load_dependencies,
+            dependencies: HashSet::default(),
+            loader_dependencies: HashMap::default(),
+            labeled_assets: HashMap::default(),
         }
     }
 
-    /// Gets the source path for this load context.
-    pub fn path(&self) -> &Path {
-        self.path
-    }
-
-    /// Returns `true` if the load context contains an asset with the specified label.
-    pub fn has_labeled_asset(&self, label: &str) -> bool {
-        self.labeled_assets.contains_key(&Some(label.to_string()))
+    /// Begins a new labeled asset load. Use the returned [`LoadContext`] to load
+    /// dependencies for the new asset and call [`LoadContext::finish`] to finalize the asset load.
+    /// When finished, make sure you call [`Self::add_labled_asset`] to add the results back to the parent
+    /// context.
+    /// Prefer [`Self::labeled_asset_scope`] when possible, which will automatically add
+    /// the labeled [`LoadContext`] back to the parent context.
+    /// [`Self::begin_labeled_asset`] exists largely to enable parallel asset loading.
+    pub fn begin_labeled_asset(&self, label: String) -> LoadContext {
+        LoadContext::new(
+            self.asset_server,
+            self.asset_path.with_label(label),
+            self.should_load_dependencies,
+        )
     }
 
-    /// Sets the primary asset loaded from the asset source.
-    pub fn set_default_asset<T: Asset>(&mut self, asset: LoadedAsset<T>) {
-        self.labeled_assets.insert(None, asset.into());
+    pub fn labeled_asset_scope<A: Asset>(
+        &mut self,
+        label: String,
+        load: impl FnOnce(&mut LoadContext) -> A,
+    ) -> Handle<A> {
+        let mut context = self.begin_labeled_asset(label.clone());
+        let asset = load(&mut context);
+        let loaded_asset = context.finish(asset, None);
+        self.add_loaded_labeled_asset(loaded_asset)
```

Comment:
This would benefit from a `trace!` scope. I think it could help for understanding loading time of individual asset. (either in tracy or just debuggability)

Reply:
I'm down for more tracing!

> load: FnOnce(&mut LoadContext) -> Result<A>. 

It would require making labeled_asset_scope result a Result too. Gotta consider the implications of that, but thats probably ok.